components/org.wso2.carbon.identity.application.authenticator.basicauth.jwt/src/main/java/org/wso2/carbon/identity/application/authenticator/basicauth/jwt/JWTBasicAuthenticator.java

```diff
@@ -0,0 +1,273 @@
+                    context.setRememberMe(true);
+                }
+            } else {
+                throw new AuthenticationFailedException("User authentication failed : Invalid signature.");
+            }
+        } else {
+            throw new AuthenticationFailedException("Invalid token");
+        }
+    }
+
+    @Override
+    public String getFriendlyName() {
+        return JWTBasicAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public String getName() {
+        return JWTBasicAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    private SignedJWT getSignedJWT(String jwtAssertion) throws AuthenticationFailedException {
+
+        String errorMessage = "No Valid JWT Assertion was found.";
+        SignedJWT signedJWT;
+        if (StringUtils.isBlank(jwtAssertion)) {
+            throw new AuthenticationFailedException(errorMessage);
+        }
+
+        try {
+            signedJWT = SignedJWT.parse(jwtAssertion);
+        } catch (ParseException e) {
+            if (log.isDebugEnabled()) {
+                log.debug(e.getMessage());
+            }
+            throw new AuthenticationFailedException("Error while parsing the JWT.");
+        }
+
+        if (signedJWT == null) {
+            throw new AuthenticationFailedException(errorMessage);
+        }
+        return signedJWT;
+    }
+
+    private JWTClaimsSet getClaimSet(SignedJWT signedJWT) throws AuthenticationFailedException {
+
+        if (signedJWT == null) {
+            throw new AuthenticationFailedException("No Valid JWT Assertion was found.");
+        }
+
+        JWTClaimsSet claimsSet;
+        try {
+            claimsSet = signedJWT.getJWTClaimsSet();
+
+            if (claimsSet == null) {
+                throw new AuthenticationFailedException("Claim values are empty in the given JWT.");
+            }
+        } catch (ParseException e) {
+            String errorMsg = "Error when trying to retrieve claimsSet from the JWT.";
+            if (log.isDebugEnabled()) {
+                log.debug(errorMsg);
+            }
+            throw new AuthenticationFailedException(errorMsg);
+        }
+        return claimsSet;
+    }
+
+    private boolean isValidClaimSet(JWTClaimsSet claimsSet) throws AuthenticationFailedException {
+
+        if (StringUtils.isEmpty(claimsSet.getSubject()) || StringUtils.isEmpty(claimsSet.getIssuer()) || StringUtils
+                .isEmpty(claimsSet.getJWTID()) || claimsSet.getExpirationTime() == null) {
+            throw new AuthenticationFailedException("Invalid token : Required fields are not present in JWT.");
+        }
+
+        if (AuthJwtCache.getInstance().getValueFromCache(claimsSet.getJWTID()) != null) {
+            throw new AuthenticationFailedException("Invalid token : Possible replay attack.");
+        }
+
+        Date currentDate = new Date();
+        if (currentDate.compareTo(claimsSet.getExpirationTime()) > 0) {
```

Comment:
Don't we need to consider timestamp skew?

Reply:
Ideally we should. Will fix